### PR TITLE
Implement numpy.gradient

### PR DESCRIPTION
**Pull request by clemisch**

### CODE REVIEW

**Review by mattjj — COMMENTED** [review]
This looks great!

Can you add a test? Anything not under continuous integration tests will eventually break!

**Comment by clemisch** [comment]
I implemented a test! Please let me know if I can improve something. 

I'm not so happy with `a_grad = [gradient_along_axis(a, ax) for ax in axis]`, but since `axis` is usually small this should be fine?

**Comment by mattjj** [comment]
That reasoning makes sense to me. The alternative would be to use `lax.scan`. But I don't think it's necessary. A user who wants to compute the gradient along many axes could write their own scan.

However, I just had a thought: could you write this in terms of a convolution, i.e. a call to `lax.conv` or `np.convolve` if we have it? I'm not sure but that might be faster on accelerators. @hawkinsp any suggestions on the implementation front?

**Comment by clemisch** [comment]
Thank you for your feedback!

I've also thought of convolution but struggled with applying it (within the jax framework) on arbitrary axes elegantly. I saw `numpy.diff` being implemented "manually" and decided against convolution for `numpy.gradient`.

**Comment by mattjj** [comment]
No, I think there's #1831 in progress. You could try on top of that branch. (I'll take a look at whether anything's holding back #1831.)

**Comment by shoyer** [comment]
> I'm not so happy with `a_grad = [gradient_along_axis(a, ax) for ax in axis]`, but since `axis` is usually small this should be fine?

I don't think `scan` would work here, because the computations are different for each axis.

> I've also thought of convolution but struggled with applying it (within the jax framework) on arbitrary axes elegantly. I saw `numpy.diff` being implemented "manually" and decided against convolution for `numpy.gradient`.

+1 for doing this the dumb way with slicing/addition/concatenation.

The gradient calculation is convolution with the kernel `[-0.5, 0, 0.5]`, which means you waste some compute multiplying/adding by zero. The kernel isn't big enough for the matmul inside XLA's conv to give any meaningful speedup, and conversely, XLA is pretty good at optimizing away intermediates arrays created by slicing.

My microbenchmark shows that the slicing version is 10-30% faster on both CPU and GPU:
https://gist.github.com/shoyer/e60484053e82ad1fc16955752b46333b

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -873,6 +873,49 @@ def diff(a, n=1, axis=-1,):
...
+@partial(jit, static_argnums=1)
+def _gradient(a, axis):
+  def gradient_along_axis(a, axis):
+    a_swap = swapaxes(a, 0, axis)
+    a_grad = concatenate((
+      (a_swap[1]  - a_swap[0])[newaxis],
+      (a_swap[2:] - a_swap[:-2]) * 0.5,
+      (a_swap[-1] - a_swap[-2])[newaxis]
+      ), axis=0)
+    return swapaxes(a_grad, 0, axis)
```
Comment (shoyer) [inline jax/numpy/lax_numpy.py]: This is totally fine, but might find it a little cleaner to make use of `jax.lax.slice_in_dim` here instead of using `swapaxes`, i.e.,
```python
    sliced = functools.partial(jax.lax.slice_in_dim, a, axis=axis)
    a_grad = concatenate((
        sliced(1, 2) - sliced(0, 1),
        (sliced(2, None) - sliced(None, -2)) * 0.5,
        sliced(-1, None) - sliced(-2, -1),
    ), axis)
```
I haven't measured this, but it might also result in slightly faster code from XLA.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -873,6 +873,49 @@ def diff(a, n=1, axis=-1,):
...
+
+  if len(axis) == 1:
+    a_grad = a_grad[0]
+
+  return a_grad
+
+
+@_wraps(onp.gradient)
+def gradient(a, *args, **kwargs):
+  axis = kwargs.pop("axis", None)
```
Comment (shoyer) [inline jax/numpy/lax_numpy.py]: we are _this close_ to dropping Python 2 support, which would let you use a real keyword argument here :)

Comment (clemisch) [reply jax/numpy/lax_numpy.py]: TBH I don't completely understand what you mean :smile:

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (clemisch) [reply jax/numpy/lax_numpy.py]: Thank you for that feedback; I was not aware of `jax.lax.slice_in_dim` and this part definitely looks cleaner like this. 

Maybe I am missing something, but `jax.lax.slice_in_dim` neither takes negative indices nor `None`s. Either I plug in that functionality into `jax.lax.slice_in_dim` or this part here gets more messy. I guess the former would not hurt?

Comment (clemisch) [reply jax/numpy/lax_numpy.py]: I've implemented support for negative and `None` indices in `slice_in_dim` and changed `_gradient_along_axis` to your suggestion. 

Performance is indeed better! See this self-contained example: https://gist.github.com/clemisch/f1462d08591eceeb377b31de4eaa2b9e

Comment (shoyer) [reply jax/numpy/lax_numpy.py]: Oh, awesome! Yes, this is my fault for writing code without testing it 😀

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -2558,6 +2558,26 @@ def testPrecision(self):
...
+         "dtype": dtype, "rng_factory": rng_factory}
+        for shape in [(10,), (10, 15), (10, 15, 20)]
+        for dtype in inexact_dtypes
+        for rng_factory in [jtu.rand_default]))
+  def testGradient(self, shape, dtype, rng_factory):
+    rng = rng_factory()
+    args_maker = self._GetArgsMaker(rng, [shape], [dtype])
+    ndim = len(shape)
+    for num_axes in range(ndim):
+      for axis in itertools.combinations(range(ndim), num_axes):
```
Comment (shoyer) [inline tests/lax_numpy_test.py]: It would be slightly preferable to move these loops into the `named_parameters` comprehension above. That should make test cases point more directly to failing tests if/when something goes wrong.

(Control flow inside test cases can be a little challenging to debug, because you usually only have access to the failing test message which doesn't tell you at which iteration the test failed.)

**Review by shoyer — APPROVED** [review]

**Comment by mattjj** [comment]
@shoyer thanks for figuring that out! Looks like my idea was misguided, but you set us straight. Hope I didn't cost us too much time here.

File: jax/lax/lax.py [hunk 4]
```diff
@@ -1320,6 +1320,17 @@ def slice_in_dim(operand, start_index, limit_index, stride=1, axis=0):
   limit_indices = list(operand.shape)
   strides = [1] * operand.ndim
 
+  # translate `None`
```
Comment (mattjj) [inline jax/lax/lax.py]: I'm wary of changing the behavior of `lax.slice_in_dim` here just for implementing a single numpy function. An alternative is just to write a helper function in the body of `_gradient` with this behavior.

Can you make the change in `_gradient` instead, basically adding a helper function instead of using `partial(lax.slice_in_dim, ...)` there?

Comment (shoyer) [reply jax/lax/lax.py]: One typical pattern is to write a helper function that leverages normal array indexing:
```
def index_along(operand, index, axis=0):
  indices = [slice(None)] * operand.ndim
  indices[axis] = index
  return operand[tuple(indices)]
```

It would be kind of nice to put something like this into JAX somewhere, though I'm not entirely sure where. For now, a little helper function here seems fine.

Maybe I'll try to push it into `np.take` upstream first?

Comment (mattjj) [reply jax/lax/lax.py]: I realize now I'm reviving a discussion in a different thread.

@shoyer do you agree with my instinct to resist changing `lax.slice_in_dim`? Or is that dumb because it's a helper function anyway and the new extended behavior is pretty reasonable?

Comment (mattjj) [reply jax/lax/lax.py]: I'd be more confident in changing `lax.slice_in_dim` if we could also run our google-internal tests.

But I just did a code search and actually nothing currently calls `lax.slice_in_dim` except our code in lax_numpy.py.

@clemisch I changed my mind about this thread. Let's just go with the PR as-is!

**Comment by mattjj** [comment]
Thanks for this contribution, and for iterating with @shoyer's useful comments and my less-than-useful ones!

**Comment by clemisch** [comment]
Thank you for all the feedback!
